contents/csharp/CSAsyncAwait.md

```diff
@@ -0,0 +1,322 @@
+```
+
+<box type="info">
+
+Returning a `Task` is usually favourable to returning `void`, since a `Task` is awaitable while `void` is not. However, there are situations where `void` is the preferred return type. The primary use case is event handlers, such as the Button Click event handler in the very first example.
+
+</box>
+
+There is a bit of a disconnect between the return type of the `async` method and the type of the value being returned, which takes some getting used to. More information on this design be found on [Stephen Cleary's blog post.](https://blog.stephencleary.com/2011/09/async-ctp-why-do-keywords-work-that-way.html)
+
+## Control Flow
+
+Let's take a look at an example to see the control flow of an asynchronous method as a whole:
+
+![Control Flow](async-await-control-flow.png)
+
+1. The `PerformTaskAsync()` method is called by an external method, which we'll call the `CallingMethod()`. Let the thread making this call be denoted as the _"main thread"_, drawn in blue lines.
+2. The method makes a call to another `async` method, `GetAnswerAsync()`.
+3. `GetAnswerAsync()` will return a `Task<int>` object while it continues in the background.
+4. The method continues its execution. This is performed _synchronously_ on the main thread.
+5. The execution hits an `await` statement. Since the `answerTask` has yet to finish, the main thread suspends the method and `return`s to the `CallingMethod`.
+6. Eventually, the `GetAnswerAsync` method finishes. The execution of the remainder of the `PerformTaskAsync` method is invoked on some thread, let's call it the _"secondary thread"_, drawn in red lines.
+7. The remainder of the method gets executed on the secondary thread.
+
+<box type="info">
+
+As mentioned before, details on which thread executes the remainder of the method can be found on [Stephen Cleary's blog post](https://blog.stephencleary.com/2012/02/async-and-await.html).
+
+</box>
+
+<box type="info">
+
+Notice how the execution of the method is synchronous until it reaches the `await` statement. This means that if the `async` method does not contain any `await` statements, the entire method will be executed synchronously. The compiler will provide a warning in such a case:
+
+```csharp
+async Task<int> NoAwaitTask()
+{
+    // Compiler warning:
+    // This async method lacks 'await' operators and will run synchronously.
+    // Consider using the 'await' operator to await non-blocking API calls,
+    // or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+
+    return 42;
+}
+```
+
+</box>
+
+## Disadvantages Of Async/Await
+
+While the use of `async` and `await` can provide the wonderful benefits of responsiveness while keeping the code readable, there are some disadvantages that come with their use.
+
+### Potential Pitfall: Deadlock
+
+It takes some time to understand the correct usage of `async` and `await`. If you're not careful, you could end up with a deadlock situation. Consider the following example:

```

Comment:
I kind of got the idea after reading the article provided. Do you think having a similar diagram from the control flow section could help illustrate the idea here?

Reply:
That might be a good idea.

I think I should probably briefly mention (possibly in the `async` section)
- Why the completed task needs to wait for the UI thread specifically (and not just any thread in the thread pool)
- How the execution resumes _at the first opportunity_ the thread is available. (Since that availability never comes, the execution can't resume)